### PR TITLE
improve pallet hooks docs

### PR DESCRIPTION
**Pull request by kianenigma**
subset of https://github.com/paritytech/substrate/pull/14279, making it ready earlier for PBA. 
Companion to #13987 

@liamaharon possibly you want to add more about migrations? 

Screenshot 2023-07-17 at 11 47 02

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3188977

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3188976

File: frame/support/src/traits/hooks.rs [hunk 1]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
+///
+/// For all hooks, except [`OnIdle`] the order of execution is derived from how the pallets are
+/// ordered in [`crate::construct_runtime`].
+///
+/// ## Summary
+///
+/// In short, the following diagram shows the
 pub trait Hooks<BlockNumber> {
-	/// The block is being finalized. Implement to have something happen.
+	/// Block initialization hook. This is called at the very beginning of the block execution
```
Comment (sam0x17) [inline frame/support/src/traits/hooks.rs:220]:
```suggestion
	/// Block initialization hook. This is called at the very beginning of block execution.
```

File: frame/support/src/traits/hooks.rs [hunk 2]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
+	/// 1. is called after [`on_initialize`] of all pallets is done executing,
+	/// 2. execution on *each block* is not guaranteed
+	///
+	/// The latter point is usually correlated with the block resources being available, for example
+	/// `Weight` leftover.
+	fn poll(_n: BlockNumber) -> Weight {
+		Weight::zero()
+	}
+
+	/// Block finalization hook. This is called at the very beginning of the block execution
```
Comment (sam0x17) [inline frame/support/src/traits/hooks.rs:248]:
```suggestion
	/// Block finalization hook. This is called at the very end of block execution.
```

File: frame/support/src/traits/hooks.rs [hunk 3]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
+	///
+	/// Note that this has nothing to do with finality in the a *consensus"* jargon.
```
Comment (sam0x17) [inline frame/support/src/traits/hooks.rs:250]:
```suggestion
	/// Note that this has nothing to do with finality in the "consensus" sense.
```

**Review by sam0x17 — APPROVED** [review]
a few nits, one that was substantive

**Comment by kianenigma** [comment]
bot fmt

**Comment by kianenigma** [comment]
bot rebase

File: frame/support/src/traits/hooks.rs [hunk 4]
```diff
@@ -336,42 +334,41 @@ pub trait Hooks<BlockNumber> {
...
 	fn integrity_test() {}
 }
 
-/// A trait to define the build function of a genesis config for both runtime and pallets.
-///
-/// Replaces deprecated [`GenesisBuild<T,I>`].
-pub trait BuildGenesisConfig: Default + sp_runtime::traits::MaybeSerializeDeserialize {
-	/// The build function puts initial `GenesisConfig` keys/values pairs into the storage.
-	fn build(&self);
-}
```
Comment (thiolliere) [inline frame/support/src/traits/hooks.rs:368]: is it removed by mistake ?

File: frame/support/src/traits/hooks.rs [hunk 5]
```diff
@@ -336,42 +334,41 @@ pub trait Hooks<BlockNumber> {
...
-
 /// A trait to define the build function of a genesis config, T and I are placeholder for pallet
 /// trait and pallet instance.
-#[deprecated(
-	note = "GenesisBuild is planned to be removed in December 2023. Use BuildGenesisConfig instead of it."
-)]
```
Comment (thiolliere) [inline frame/support/src/traits/hooks.rs:374]: this too ?

File: frame/support/src/traits/hooks.rs [hunk 6]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
+/// - [`IntegrityTest`]
+///
+/// ## Ordering
+///
+/// For all hooks, except [`OnIdle`] the order of execution is derived from how the pallets are
+/// ordered in [`crate::construct_runtime`].
+///
+/// ## Summary
+///
+/// In short, the following diagram shows the
```
Comment (thiolliere) [inline frame/support/src/traits/hooks.rs:218]: the end of the sentence is missing?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3197988

File: frame/support/src/traits/hooks.rs [hunk 7]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
 pub trait Hooks<BlockNumber> {
-	/// The block is being finalized. Implement to have something happen.
+	/// Block initialization hook. This is called at the very beginning of block execution.
+	///
+	/// Must return the non-negotiable weight of both itself, and whatever [`on_finalize`] wishes to
+	/// consumed.
```
Comment (juangirini) [inline frame/support/src/traits/hooks.rs:223]:
```suggestion
	/// Must return the non-negotiable weight of both itself and whatever [`on_finalize`] wishes to
	/// consume.
```

File: frame/support/src/traits/hooks.rs [hunk 8]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
+	}
+
+	/// Block finalization hook. This is called at the very end of block execution.
+	///
+	/// Note that this has nothing to do with finality in the "consensus" sense.
+	///
+	/// Note that the non-negotiable weight for this has must have already been returned by
+	/// [`OnInitialize`]. It usage along is not permitted.
+	///
+	/// Similar to [`OnInitialize`] it should only be used when execution is absolutely necessary.
```
Comment (juangirini) [inline frame/support/src/traits/hooks.rs:255]:
```suggestion
	/// Note that the non-negotiable weight for this has must have already been returned by
	/// [`on_initialize`]. It usage along is not permitted.
	///
	/// Similar to [`on_initialize`] it should only be used when execution is absolutely necessary.
```

File: frame/support/src/traits/hooks.rs (hunk 4 above)
Comment (kianenigma) [reply frame/support/src/traits/hooks.rs:368]: No I must have been out of sync with master 🙈

Comment (kianenigma) [reply frame/support/src/traits/hooks.rs:368]: will fix soon.

File: frame/support/src/traits/hooks.rs [hunk 9]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
 
-	/// This will be run when the block is being finalized (before `on_finalize`).
+	/// Hook to consume a block's idle time. This will run when the block is being finalized (before
+	/// [`on_finalize`]).
 	///
-	/// Implement to have something happen using the remaining weight. Will not fire if the
-	/// remaining weight is 0.
+	/// Given that all dispatchables are already executed and noted, and the weight for
+	/// [`on_finalize`], which comes next, is also already accounted for (via `on_initialize`), this
+	/// hook consumes anything that is leftover.
```
Comment (juangirini) [inline frame/support/src/traits/hooks.rs:263]: I get confused with many commas, maybe something like this?

```suggestion
	/// Given that all dispatchables are already executed and noted (and the weight for
	/// [`on_finalize`], which comes next, is also already accounted for via `on_initialize`), this
	/// hook consumes anything that is leftover.
```

File: frame/support/src/traits/hooks.rs [hunk 10]
```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
...
+	/// Hook executed when a code change (aka. a "runtime upgrade") is detected by FRAME.
 	///
-	/// Return the non-negotiable weight consumed in the block.
-	fn on_initialize(_n: BlockNumber) -> Weight {
-		Weight::zero()
-	}
-
-	/// Perform a module upgrade.
+	/// Be aware that this is called before [`on_initialize`] of any pallet, therefore a lot of the
+	/// critical storage items such as `block_number` in system pallet might not been set.
```
Comment (juangirini) [inline frame/support/src/traits/hooks.rs:280]:
```suggestion
	/// Be aware that this is called before [`on_initialize`] of any pallet; therefore, a lot of the
	/// critical storage items such as `block_number` in system pallet might have not been set.
```

File: frame/support/src/traits/hooks.rs [hunk 11]
```diff
@@ -336,42 +334,41 @@ pub trait Hooks<BlockNumber> {
...
+	/// [`sp-io`], namely [`sp_io::offchain`]. This includes exotic operations such as HTTP calls
+	/// that are not really possible in the rest of the runtime code.
 	///
-	/// Implement this and use any of the `Offchain` `sp_io` set of APIs
-	/// to perform off-chain computations, calls and submit transactions
-	/// with results to trigger any on-chain changes.
-	/// Any state alterations are lost and are not persisted.
+	/// The execution of this hook is entirely optional and is left at the discretion of the node
+	/// side software, and its configuration. In a normal substrate-cli, look for the CLI flags
+	/// related to offchain-workers to learn more.
```
Comment (juangirini) [inline frame/support/src/traits/hooks.rs:351]:
```suggestion
	/// The execution of this hook is entirely optional and is left at the discretion of the
	/// node-side software and its configuration. In a normal substrate-cli, look for the CLI
	/// flags related to offchain-workers to learn more.
```

File: frame/support/src/traits/hooks.rs [hunk 12]
```diff
@@ -336,42 +334,41 @@ pub trait Hooks<BlockNumber> {
...
-	/// Run integrity test.
+	/// Check the integrity of this pallet's configuration.
+	///
+	/// Any code located in this hook is placed in an auto-generated test, and generated as a part
+	/// of [`construct_runtime`]'s expansion. Look for a test case with a name along the lines of:
+	/// `__construct_runtime_integrity_test`.
 	///
-	/// The test is not executed in a externalities provided environment.
+	/// This hook is suitable location to ensure value values/types provided to the `Config` trait
+	/// of the pallet can be tested for correctness. For example, if two `type Foo: Get<u32>` and
```
Comment (juangirini) [inline frame/support/src/traits/hooks.rs:361]:
```suggestion
	/// This hook is the location where the values/types provided to the `Config` trait
	/// of the pallet can be tested for correctness. For example, if two `type Foo: Get<u32>` and
```

**Review by juangirini — APPROVED** [review]
Just some nits

File: frame/support/src/traits/hooks.rs (hunk 6 above)
Comment (kianenigma) [reply frame/support/src/traits/hooks.rs:218]: Adding a mermaid diagram now.

File: frame/support/src/traits/hooks.rs [hunk 13]
```diff
@@ -239,64 +190,137 @@ impl OnRuntimeUpgrade for Tuple {
...
+	fn on_initialize(_n: BlockNumber) -> Weight {
+		Weight::zero()
+	}
+
+	/// Block finalization hook. This is called at the very end of block execution.
+	///
+	/// Note that this has nothing to do with finality in the "consensus" sense.
+	///
+	/// Note that the non-negotiable weight for this has must have already been returned by
+	/// [`Hooks::on_initialize`]. It usage along is not permitted.
```
Comment (gpestana) [inline frame/support/src/traits/hooks.rs:275]: > It usage along is not permitted.


I don't get what this means.

**Review by gpestana — APPROVED** [review]

File: frame/support/src/traits/hooks.rs (hunk 13 above)
Comment (kianenigma) [reply frame/support/src/traits/hooks.rs]: usage "_alone_" 🙈 I meant to say you cannot use this without `on_initialize`.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**